cdc/sinkv2/eventsink/mq/mq_sink.go

```diff
@@ -14,24 +14,122 @@
 )
 
+// newProducerImpl specifies the build method for the producer.
+var newProducerImpl producer.NewProducerFunc = producer.NewMockProducer
+
 // Assert EventSink[E event.TableEvent] implementation
-var _ eventsink.EventSink[*model.RowChangedEvent] = (*Sink)(nil)
+var _ eventsink.EventSink[*model.RowChangedEvent] = (*sink)(nil)
+
+// sink is the mq sink.
+// It will send the events to the MQ system.
+type sink struct {
+	// id indicates this sink belongs to which processor(changefeed).
+	id model.ChangeFeedID
+	// role indicates this sink used for what.
+	role util.Role
+	// protocol indicates the protocol used by this sink.
+	protocol config.Protocol
+
+	worker *worker
+	// eventRouter used to route events to the right topic and partition.
+	eventRouter *dispatcher.EventRouter
+	// topicManager used to manage topics.
+	// It is also responsible for creating topics.
+	topicManager manager.TopicManager
+
+	// encoderBuilder builds encoder for the sink.
+	encoderBuilder codec.EncoderBuilder
+}
+
+func newSink(ctx context.Context,
+	producer producer.Producer,
+	topicManager manager.TopicManager,
+	eventRouter *dispatcher.EventRouter,
+	encoderConfig *codec.Config,
+	errCh chan error,
+) (*sink, error) {
+	changefeedID := contextutil.ChangefeedIDFromCtx(ctx)
+	role := contextutil.RoleFromCtx(ctx)
 
-// Sink is the mq sink.
-type Sink struct{}
+	encoderBuilder, err := codec.NewEventBatchEncoderBuilder(ctx, encoderConfig)
+	if err != nil {
+		return nil, cerror.WrapError(cerror.ErrKafkaInvalidConfig, err)
+	}
+	encoder := encoderBuilder.Build()
+
+	w := newWorker(changefeedID, encoder, producer)
+
+	s := &sink{
+		id:             changefeedID,
+		role:           role,
+		protocol:       encoderConfig.Protocol(),
+		worker:         w,
+		eventRouter:    eventRouter,
+		topicManager:   topicManager,
+		encoderBuilder: encoderBuilder,
+	}
+
+	// Spawn a goroutine to send messages by the worker.
+	go func() {
+		if err := s.worker.run(ctx); err != nil && errors.Cause(err) != context.Canceled {
+			select {
+			case <-ctx.Done():
+				return
+			case errCh <- err:
+			default:
+				log.Error("error channel is full", zap.Error(err),
+					zap.String("namespace", changefeedID.Namespace),
+					zap.String("changefeed", changefeedID.ID),
+					zap.Any("role", s.role))
+			}
+		}
+	}()
+
+	return s, nil
+}
 
 // WriteEvents writes events to the sink.
-func (s *Sink) WriteEvents(rows ...*eventsink.RowChangeCallbackableEvent) {
-	// TODO implement me
-	panic("implement me")
+// This is an asynchronously and thread-safe method.
+func (s *sink) WriteEvents(rows ...*eventsink.RowChangeCallbackableEvent) error {
+	for _, row := range rows {
+		topic := s.eventRouter.GetTopicForRowChange(row.Event)
+		partitionNum, err := s.topicManager.GetPartitionNum(topic)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		partition := s.eventRouter.GetPartitionForRowChange(row.Event, partitionNum)
+		// This never be blocked.
+		s.worker.msgChan.In() <- mqEvent{
```

Comment:
Looks like it might be blocked forever, could you select it among with a context?

Reply:
This is an unbound chan, so it shouldn't be block.